core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerde.java

```diff
@@ -165,161 +136,1033 @@ private static MarkerSet getSerializedPageMarkerSet(Slice serializedPage)
+            while (bytesRemaining > 0) {
+                ensureReadable(min(Long.BYTES, bytesRemaining));
+                int bytesToRead = min(bytesRemaining, buffer.available());
+                int bytesRead = buffer.read(destination, destinationIndex, bytesToRead);
+                if (bytesRead == -1) {
+                    break;
+                }
+                bytesRemaining -= bytesRead;
+                destinationIndex += bytesRead;
+            }
+            return length - bytesRemaining;
+        }
+
+        @Override
+        public void readBytes(byte[] destination, int destinationIndex, int length)
+        {
+            ReadBuffer buffer = buffers[0];
+            int bytesRemaining = length;
+            while (bytesRemaining > 0) {
+                ensureReadable(min(Long.BYTES, bytesRemaining));
+                int bytesToRead = min(bytesRemaining, buffer.available());
+                buffer.readBytes(destination, destinationIndex, bytesToRead);
+                bytesRemaining -= bytesToRead;
+                destinationIndex += bytesToRead;
             }
         }
 
-        private DynamicSliceOutput acquireSliceOutput(int estimatedSize)
+        @Override
+        public void readBytes(Slice destination, int destinationIndex, int length)
         {
-            checkNotClosed();
-            if (sliceOutput != null && sliceOutput.writableBytes() >= estimatedSize) {
-                DynamicSliceOutput result = this.sliceOutput;
-                this.sliceOutput = null;
-                return result;
+            ReadBuffer buffer = buffers[0];
+            int bytesRemaining = length;
+            while (bytesRemaining > 0) {
+                ensureReadable(min(Long.BYTES, bytesRemaining));
+                int bytesToRead = min(bytesRemaining, buffer.available());
+                buffer.readBytes(destination, destinationIndex, bytesToRead);
+                bytesRemaining -= bytesToRead;
+                destinationIndex += bytesToRead;
             }
-            this.sliceOutput = null; // Clear any existing slice output that might be smaller than the request
-            return new DynamicSliceOutput(estimatedSize);
         }
 
-        private void releaseSliceOutput(DynamicSliceOutput sliceOutput)
+        private void ensureReadable(int bytes)
         {
-            if (closed) {
+            if (buffers[0].available() >= bytes) {
                 return;
             }
-            sliceOutput.reset();
-            if (sliceOutput.writableBytes() <= MAX_BUFFER_RETAINED_SIZE) {
-                this.sliceOutput = sliceOutput;
+            decrypt();
+            decompress();
+        }
+
+        private void decrypt()
+        {
+            if (this.encryptionKey.isEmpty()) {
+                return;
             }
+
+            ReadBuffer source = buffers[buffers.length - 1];
+            ReadBuffer sink = buffers[buffers.length - 2];
+            int bytesPreserved = sink.rollOver();
```

Comment:
do you need to return `bytesPreserved`. Isn't it just sink.available()?

Reply:
The idea is to preserve the bytes that haven't been read (for example if there are less than 8 bytes available and a single long is being read). This method returns the number of bytes preserved to offset the write index. It felt like it might be easier to follow this way, but yeah, it can also be implemented as:

```
int bytesPreserved = sink.available()
sink.rollOver();
```

I don't have a strong opinion here.

Reply:
neither do i